cpp/src/io/json/parser_features.cpp

```diff
@@ -16,14 +16,194 @@
 
 #include "nested_json.hpp"
 
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
 #include <cudf/detail/utilities/visitor_overload.hpp>
+#include <cudf/dictionary/dictionary_factories.hpp>
+#include <cudf/io/json.hpp>
+#include <cudf/strings/string_view.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
 
 #include <optional>
 #include <string>
 #include <vector>
 
+namespace cudf::io {
+namespace {
+bool validate_column_order(schema_element const& types)
+{
+  // For struct types, check if column_order size matches child_types size and all elements in
+  // column_order are in child_types, in child_types, call this function recursively.
+  // For list types, check if child_types size is 1 and call this function recursively.
+  if (types.type.id() == type_id::STRUCT) {
+    if (types.column_order.has_value()) {
+      if (types.column_order.value().size() != types.child_types.size()) { return false; }
+      for (auto const& column_name : types.column_order.value()) {
+        auto it = types.child_types.find(column_name);
+        if (it == types.child_types.end()) { return false; }
+        if (it->second.type.id() == type_id::STRUCT or it->second.type.id() == type_id::LIST) {
+          if (!validate_column_order(it->second)) { return false; }
+        }
+      }
+    }
+  } else if (types.type.id() == type_id::LIST) {
+    if (types.child_types.size() != 1) { return false; }
+    auto it = types.child_types.begin();
+    if (it->second.type.id() == type_id::STRUCT or it->second.type.id() == type_id::LIST) {
+      if (!validate_column_order(it->second)) { return false; }
+    }
```

Comment:
Here we still didn't check for column name existence.

Reply:
Should be
```
  auto it = types.child_types.find(column_name);
  if (it == types.child_types.end()) { return false; }
```

Reply:
This is not required.  For non-nested type, there is no column_order.
Nested types are taken care by STRUCT, and LIST if cases.